server/src/main/java/com/linecorp/centraldogma/server/metadata/RepositoryMetadataDeserializer.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2024 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.centraldogma.server.metadata;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.linecorp.centraldogma.server.metadata.PerRolePermissions.READ_WRITE;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import javax.annotation.Nullable;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.centraldogma.common.RepositoryRole;
+import com.linecorp.centraldogma.internal.Jackson;
+import com.linecorp.centraldogma.server.QuotaConfig;
+
+final class RepositoryMetadataDeserializer extends StdDeserializer<RepositoryMetadata> {
+
+    private static final long serialVersionUID = 1173216371065909688L;
+
+    private static final TypeReference<Map<String, Collection<Permission>>> PER_PERMISSIONS_TYPE =
+            new TypeReference<Map<String, Collection<Permission>>>() {};
+
+    RepositoryMetadataDeserializer() {
+        super(RepositoryMetadata.class);
+    }
+
+    @Override
+    public RepositoryMetadata deserialize(JsonParser p, DeserializationContext ctxt)
+            throws IOException {
+        final JsonNode jsonNode = p.readValueAsTree();
+        final String name = jsonNode.get("name").textValue();
+        final JsonNode perRolePermissionsNode = jsonNode.get("perRolePermissions");
```

Comment:
Note: I understood that if both `perRolePermissions` and `roles` is in the json node, then `perRolePermissions` (the legacy format) has higher priority.

Reply:
>  I understood that if both perRolePermissions and roles is in the json node

They cannot be. Let me ass an assertion logic. 😉